calyx-frontend/src/parser.rs

```diff
@@ -1321,24 +1360,123 @@ impl CalyxParser {
+                (PositionId::new(pos_num), MetadataFileId::new(file_num), LineNum::new(line_no))}
+        ))
+    }
+
+    fn position_table(
+        input: Node,
+    ) -> ParseResult<
+        impl IntoIterator<Item = (PositionId, MetadataFileId, LineNum)>,
+    > {
+        Ok(match_nodes!(input.into_children();
+                [position_header(_), position_entry(e)..] => e))
+    }
+
+    fn metadata_table(
+        input: Node,
+    ) -> ParseResult<SourceInfoResult<SourceInfoTable>> {
+        Ok(match_nodes!(input.into_children();
+            [file_table(f), position_table(p)] => SourceInfoTable::new(f, p)
+        ))
+    }
+
+    // end new metadata
+
+    fn extra_info(
+        input: Node,
+    ) -> ParseResult<(Option<String>, Option<SourceInfoTable>)> {
+        Ok(match_nodes!(input.into_children();
+                [metadata_legacy(l)] => (Some(l), None),
+                [metadata_table(t)] => {
+                    if let Ok(t) = t {
+                        (None, Some(t))
+                    } else {
+                        log::error!("{}", t.unwrap_err());
+                        (None, None)
+                    }
+                },
+                [metadata_legacy(l), metadata_table(t)] => {
+                    if let Ok(t) = t {
+                        (Some(l), Some(t))
+                    } else {
+                        log::error!("{}", t.unwrap_err());
+                        (Some(l), None)
+                    }
+                },
+                [metadata_table(t), metadata_legacy(l)] => {
+                    if let Ok(t) = t {
+                        (Some(l), Some(t))
+                    } else {
+                        log::error!("{}", t.unwrap_err());
+                        (Some(l), None)
+                    }
+                }
+        ))
+    }
+
     fn file(input: Node) -> ParseResult<ast::NamespaceDef> {
         Ok(match_nodes!(
             input.into_children();
             // There really seems to be no straightforward way to resolve this
             // duplication
-            [imports(imports), externs_and_comps(mixed), metadata(m), EOI(_)] => {
+            [imports(imports), externs_and_comps(mixed), extra_info(info), EOI(_)] => {
+                let (mut legacy_metadata, metadata) = info;
+                // remove empty metadata strings
+                if let Some(m) = &legacy_metadata {
+                    if m.is_empty() {
+                        legacy_metadata = None;
+                    }
+                }
+
                 let mut namespace =
                     ast::NamespaceDef {
                         imports,
                         components: Vec::new(),
                         externs: Vec::new(),
-                        metadata: if m != *"" { Some(m) } else { None }
+                        metadata: legacy_metadata,
+                        file_info_table: metadata
```

Comment:
since you called it `SourceInfoTable`, change the name of this field to match (`source_info_table`)